File: internal/jsonhandler/jsonhandler.go
package jsonhandler

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

var (
	ErrFileNotFound   = errors.New("FILE_NOT_FOUND")
	ErrInvalidJSON    = errors.New("INVALID_JSON")
	ErrFileReadError  = errors.New("FILE_READ_ERROR")
	ErrFileWriteError = errors.New("FILE_WRITE_ERROR")
	ErrParseError     = errors.New("PARSE_ERROR")
	ErrUnknownError   = errors.New("UNKNOWN_ERROR")
)

// JSONHandler handles JSON file operations with caching support
type JSONHandler struct {
	filePath   string
	cachedData map[string]interface{}
	fileMTime  time.Time
	mutex      sync.RWMutex
}

// NewJSONHandler creates a new JSON handler for a specific file
func NewJSONHandler(filePath string) *JSONHandler {
	return &JSONHandler{
		filePath: filePath,
	}
}

// LoadJSON loads JSON data from file with optional caching
func (h *JSONHandler) LoadJSON(useCache bool) (map[string]interface{}, error) {
	h.mutex.Lock()
	defer h.mutex.Unlock()

	// Check if file exists
	fileInfo, err := os.Stat(h.filePath)
	if os.IsNotExist(err) {
		return nil, fmt.Errorf("%w: File %s not found", ErrFileNotFound, h.filePath)
	}
	if err != nil {
		return nil, fmt.Errorf("%w: Failed to stat %s: %v", ErrFileReadError, h.filePath, err)
	}

	currentMTime := fileInfo.ModTime()

	// Use cache if available and file hasn't changed
	if useCache && h.cachedData != nil && h.fileMTime.Equal(currentMTime) {
		return h.cachedData, nil
	}

	// Read and parse file
	data, err := os.ReadFile(h.filePath)
	if err != nil {
		return nil, fmt.Errorf("%w: Failed to read %s: %v", ErrFileReadError, h.filePath, err)
	}

	var jsonData map[string]interface{}
	if err := json.Unmarshal(data, &jsonData); err != nil {
		return nil, fmt.Errorf("%w: File %s contains invalid JSON: %v", ErrInvalidJSON, h.filePath, err)
	}

	// Update cache
	if useCache {
		h.cachedData = jsonData
		h.fileMTime = currentMTime
	}

	return jsonData, nil
}

// SaveJSON saves JSON data to file with atomic write
func (h *JSONHandler) SaveJSON(data map[string]interface{}, indent int) error {
	h.mutex.Lock()
	defer h.mutex.Unlock()

	// Use atomic write - write to temp file then rename
	dir := filepath.Dir(h.filePath)
	tempFile, err := os.CreateTemp(dir, "*.tmp")
	if err != nil {
		return fmt.Errorf("%w: Failed to create temp file: %v", ErrFileWriteError, err)
	}
	tempPath := tempFile.Name()
	
	defer func() {
		tempFile.Close()
		// Clean up temp file if it still exists
		os.Remove(tempPath)
	}()

	// Encode JSON with indentation
	encoder := json.NewEncoder(tempFile)
	encoder.SetIndent("", getIndentString(indent))
	encoder.SetEscapeHTML(false)
	
	if err := encoder.Encode(data); err != nil {
		return fmt.Errorf("%w: Failed to encode JSON: %v", ErrFileWriteError, err)
	}

	if err := tempFile.Close(); err != nil {
		return fmt.Errorf("%w: Failed to close temp file: %v", ErrFileWriteError, err)
	}

	// Atomic rename
	if err := os.Rename(tempPath, h.filePath); err != nil {
		return fmt.Errorf("%w: Failed to rename temp file: %v", ErrFileWriteError, err)
	}

	// Update cache
	h.cachedData = data
	if fileInfo, err := os.Stat(h.filePath); err == nil {
		h.fileMTime = fileInfo.ModTime()
	}

	return nil
}

// ValidationResult represents the result of JSON validation
type ValidationResult struct {
	Valid       bool                   `json:"valid"`
	File        string                 `json:"file"`
	Error       *ValidationError       `json:"error,omitempty"`
	ErrorType   string                 `json:"error_type,omitempty"`
	Performance *PerformanceMetrics    `json:"performance,omitempty"`
}

// ValidationError represents a JSON validation error
type ValidationError struct {
	Message string `json:"message"`
	Line    int    `json:"line"`
	Column  int    `json:"column"`
}

// PerformanceMetrics represents performance metrics
type PerformanceMetrics struct {
	ParseTime float64 `json:"parse_time"`
	FileSize  int64   `json:"file_size"`
}

// ValidateJSONSyntax validates JSON file syntax without loading into memory completely
func (h *JSONHandler) ValidateJSONSyntax() *ValidationResult {
	result := &ValidationResult{
		File: h.filePath,
	}

	// Check if file exists
	fileInfo, err := os.Stat(h.filePath)
	if os.IsNotExist(err) {
		result.Valid = false
		result.ErrorType = "FILE_NOT_FOUND"
		result.Error = &ValidationError{
			Message: fmt.Sprintf("File %s not found", h.filePath),
			Line:    0,
			Column:  0,
		}
		return result
	}

	fileSize := fileInfo.Size()

	if fileSize == 0 {
		result.Valid = false
		result.ErrorType = "PARSE_ERROR"
		result.Error = &ValidationError{
			Message: "File is empty",
			Line:    1,
			Column:  1,
		}
		return result
	}

	// Read file content
	startTime := time.Now()
	data, err := os.ReadFile(h.filePath)
	if err != nil {
		result.Valid = false
		result.ErrorType = "FILE_READ_ERROR"
		result.Error = &ValidationError{
			Message: fmt.Sprintf("Failed to read file: %v", err),
			Line:    0,
			Column:  0,
		}
		return result
	}

	// Check for empty content after reading
	if len(data) == 0 {
		result.Valid = false
		result.ErrorType = "PARSE_ERROR"
		result.Error = &ValidationError{
			Message: "File contains only whitespace",
			Line:    1,
			Column:  1,
		}
		return result
	}

	// Try to parse JSON
	var jsonData interface{}
	if err := json.Unmarshal(data, &jsonData); err != nil {
		result.Valid = false
		result.ErrorType = "PARSE_ERROR"
		
		// Try to extract line/column information from JSON error
		if jsonErr, ok := err.(*json.SyntaxError); ok {
			line, col := getLineColumn(data, jsonErr.Offset)
			result.Error = &ValidationError{
				Message: jsonErr.Error(),
				Line:    line,
				Column:  col,
			}
		} else {
			result.Error = &ValidationError{
				Message: err.Error(),
				Line:    0,
				Column:  0,
			}
		}
		return result
	}

	// Success
	parseTime := time.Since(startTime).Seconds()
	result.Valid = true
	result.Performance = &PerformanceMetrics{
		ParseTime: parseTime,
		FileSize:  fileSize,
	}

	return result
}

// ClearCache clears cached data to force reload on next access
func (h *JSONHandler) ClearCache() {
	h.mutex.Lock()
	defer h.mutex.Unlock()
	
	h.cachedData = nil
	h.fileMTime = time.Time{}
}

// FileInfo represents file information
type FileInfo struct {
	Exists       bool      `json:"exists"`
	Path         string    `json:"path"`
	SizeBytes    int64     `json:"size_bytes,omitempty"`
	ModifiedTime time.Time `json:"modified_time,omitempty"`
	IsCached     bool      `json:"is_cached"`
}

// GetFileInfo gets information about the JSON file
func (h *JSONHandler) GetFileInfo() *FileInfo {
	h.mutex.RLock()
	defer h.mutex.RUnlock()

	info := &FileInfo{
		Path:     h.filePath,
		IsCached: h.cachedData != nil,
	}

	if fileInfo, err := os.Stat(h.filePath); err == nil {
		info.Exists = true
		info.SizeBytes = fileInfo.Size()
		info.ModifiedTime = fileInfo.ModTime()
	} else {
		info.Exists = false
	}

	return info
}

// Helper function to get indent string
func getIndentString(indent int) string {
	result := ""
	for i := 0; i < indent; i++ {
		result += " "
	}
	return result
}

// Helper function to calculate line and column from byte offset
func getLineColumn(data []byte, offset int64) (int, int) {
	line := 1
	col := 1
	
	for i := int64(0); i < offset && i < int64(len(data)); i++ {
		if data[i] == '\n' {
			line++
			col = 1
		} else {
			col++
		}
	}
	
	return line, col
}

File: src/json_handler.py
"""
JSON Handler Module

Provides efficient JSON file operations with streaming support for large files.
Handles file I/O, parsing, and serialization with proper error handling.
"""

import json
import os
from pathlib import Path
from typing import Any, Dict, Optional


class JsonHandler:
    """Handles JSON file operations with streaming support for large files"""
    
    def __init__(self, file_path: str | Path):
        """
        Initialize JSON handler for a specific file
        
        Args:
            file_path: Path to the JSON file
        """
        self.file_path = Path(file_path)
        self._cached_data: Optional[Dict[str, Any]] = None
        self._file_mtime: Optional[float] = None
    
    def load_json(self, use_cache: bool = True) -> Dict[str, Any]:
        """
        Load JSON data from file with optional caching
        
        Args:
            use_cache: Whether to use cached data if available
            
        Returns:
            Parsed JSON data as dictionary
            
        Raises:
            FileNotFoundError: If file doesn't exist
            ValueError: If file contains invalid JSON
        """
        if not self.file_path.exists():
            raise FileNotFoundError(f"FILE_NOT_FOUND: File {self.file_path} not found")
        
        current_mtime = self.file_path.stat().st_mtime
        
        # Use cache if available and file hasn't changed
        if (use_cache and 
            self._cached_data is not None and 
            self._file_mtime == current_mtime):
            return self._cached_data
        
        try:
            with open(self.file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                
            # Update cache
            if use_cache:
                self._cached_data = data
                self._file_mtime = current_mtime
                
            return data
            
        except json.JSONDecodeError as e:
            raise ValueError(f"INVALID_JSON: File {self.file_path} contains invalid JSON: {e}")
        except Exception as e:
            raise ValueError(f"FILE_READ_ERROR: Failed to read {self.file_path}: {e}")
    
    def save_json(self, data: Dict[str, Any], indent: int = 2) -> None:
        """
        Save JSON data to file with atomic write
        
        Args:
            data: Dictionary to save as JSON
            indent: JSON indentation level
            
        Raises:
            ValueError: If data cannot be serialized to JSON
        """
        try:
            # Use atomic write - write to temp file then rename
            temp_path = self.file_path.with_suffix('.tmp')
            
            with open(temp_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=indent, ensure_ascii=False)
            
            # Atomic rename
            temp_path.replace(self.file_path)
            
            # Update cache
            self._cached_data = data
            self._file_mtime = self.file_path.stat().st_mtime
            
        except Exception as e:
            # Clean up temp file if it exists
            if temp_path.exists():
                temp_path.unlink()
            raise ValueError(f"FILE_WRITE_ERROR: Failed to write {self.file_path}: {e}")
    
    def validate_json_syntax(self) -> Dict[str, Any]:
        """
        Validate JSON file syntax without loading into memory completely
        
        Returns:
            Validation result dictionary with 'valid', 'error', 'file' fields
        """
        if not self.file_path.exists():
            raise FileNotFoundError(f"FILE_NOT_FOUND: File {self.file_path} not found")
        
        file_size = self.file_path.stat().st_size
        
        if file_size == 0:
            return {
                'valid': False,
                'file': str(self.file_path),
                'error_type': 'PARSE_ERROR',
                'error': {
                    'message': 'File is empty',
                    'line': 1,
                    'column': 1
                }
            }
        
        try:
            with open(self.file_path, 'r', encoding='utf-8') as f:
                content = f.read()
                
            if not content.strip():
                return {
                    'valid': False,
                    'file': str(self.file_path),
                    'error_type': 'PARSE_ERROR',
                    'error': {
                        'message': 'File contains only whitespace',
                        'line': 1,
                        'column': 1
                    }
                }
            
            # Try to parse
            json.loads(content)
            
            return {
                'valid': True,
                'file': str(self.file_path),
                'error': None
            }
            
        except json.JSONDecodeError as e:
            return {
                'valid': False,
                'file': str(self.file_path),
                'error_type': 'PARSE_ERROR',
                'error': {
                    'message': str(e),
                    'line': e.lineno,
                    'column': e.colno
                }
            }
        except Exception as e:
            return {
                'valid': False,
                'file': str(self.file_path),
                'error_type': 'UNKNOWN_ERROR',
                'error': {
                    'message': str(e),
                    'line': 0,
                    'column': 0
                }
            }
    
    def clear_cache(self) -> None:
        """Clear cached data to force reload on next access"""
        self._cached_data = None
        self._file_mtime = None
    
    def get_file_info(self) -> Dict[str, Any]:
        """
        Get information about the JSON file
        
        Returns:
            Dictionary with file metadata
        """
        if not self.file_path.exists():
            return {
                'exists': False,
                'path': str(self.file_path)
            }
        
        stat = self.file_path.stat()
        return {
            'exists': True,
            'path': str(self.file_path),
            'size_bytes': stat.st_size,
            'modified_time': stat.st_mtime,
            'is_cached': self._cached_data is not None
        }